### PR TITLE
1145 observed data read only

### LINKED ISSUE #1145: After import: observed data may not be modified anymore (except for the meta data)
@Yuri05 commented on [Tue Jun 29 2021](https://github.com/Open-Systems-Pharmacology/PK-Sim/issues/1820)

It should not be possible to edit observed **data** in PK-Sim (**meta data** can be edited as before)
If data needs to be changed: the new "Update" functionality must be used instead.

grafik

### PR DESCRIPTION
**Pull request by georgeDaskalakis**
closes #1145

### CODE REVIEW

File: src/OSPSuite.Presentation/Presenters/ObservedData/DataRepositoryDataPresenter.cs [hunk 1]
```diff
@@ -97,7 +60,7 @@ public void AddData(int rowIndex)
       public IEnumerable<string> GetCellValidationErrorMessages(int rowIndex, int columnIndex, string newValue)
```
Comment (msevestre) [inline src/OSPSuite.Presentation/Presenters/ObservedData/DataRepositoryDataPresenter.cs:60]: Why do we need this? This is for validation of new value and we can't edit. Remove

Comment (msevestre) [reply src/OSPSuite.Presentation/Presenters/ObservedData/DataRepositoryDataPresenter.cs:60]: unless this is used for something else

File: src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs [hunk 2]
```diff
@@ -1,79 +1,37 @@
...
-using OSPSuite.UI.Controls;
 using OSPSuite.UI.Services;
 
 namespace OSPSuite.UI.Views.ObservedData
 {
-   public partial class DataRepositoryDataView : BaseDataRepositoryDataView<IDataRepositoryDataView, IDataRepositoryDataPresenter>, IDataRepositoryDataView
+   public partial class DataRepositoryDataView : BaseDataRepositoryDataView<IDataRepositoryDataView, IDataRepositoryDataPresenter>,
+      IDataRepositoryDataView
    {
       protected readonly GridViewColumnUnitsMenuBinder<int> _columnUnitsMenuBinder;
```
Comment (msevestre) [inline src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs:14]: It should be removed too

File: src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs [hunk 3]
```diff
@@ -1,79 +1,37 @@
...
-         }
-         gridView.OptionsView.ShowIndicator = false;
-
-         _editable = false;
       }
 
       public override void InitializeBinding()
       {
          base.InitializeBinding();
          gridView.ValidatingEditor += (o, e) => OnEvent(gridEditorValidating, e);
```
Comment (msevestre) [inline src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs:27]: no editor. The view is readonly

File: src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs [hunk 4]
```diff
@@ -1,79 +1,37 @@
...
-         gridView.CellValueChanged += (o, e) => OnEvent(gridViewOnCellValueChanged, e);
-         _removeButtonRepository.ButtonClick += (o, e) => OnEvent(_presenter.RemoveData, gridView.GetDataSourceRowIndex(gridView.FocusedRowHandle));
-         btnAddData.Click += (o, e) => OnEvent(_presenter.AddRow);
       }
 
       public override void BindTo(DataTable dataTable)
       {
          _dataTable = dataTable;
          base.BindTo(dataTable);
          _columnUnitsMenuBinder.BindTo(_presenter);
```
Comment (msevestre) [inline src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs:34]: Do we need this?

File: src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs [hunk 5]
```diff
@@ -88,32 +46,5 @@ private void gridEditorValidating(BaseContainerValidateEditorEventArgs e)
          e.Valid = false;
          e.ErrorText = validationFailureMessages.First();
```
Comment (msevestre) [inline src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs:47]: that should go too I believe

**Review by msevestre — COMMENTED** [review]
I think there is more to remove

File: src/OSPSuite.Presentation/Presenters/ObservedData/DataRepositoryDataPresenter.cs (hunk 1 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Presentation/Presenters/ObservedData/DataRepositoryDataPresenter.cs:60]: true, this is not needed anymore

File: src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs (hunk 3 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs:27]: removed

**Comment by georgeDaskalakis** [comment]
@msevestre @Yuri05 @PavelBal actually I just realized that the context menu for observed data still say "Edit..." for the entry that opens the editor. I would rename this to "View..." if everybody agrees.

File: src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs (hunk 4 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs:28]: @msevestre are you refering to the ValidatingEditor? If yes, then I agree it is not necessary. I am checking a bit if everything still works when removing all the code

**Comment by Yuri05** [comment]
Well, the user still can edit the meta data, so I would keep it as edit

File: src/OSPSuite.Presentation/Presenters/ObservedData/DataRepositoryDataPresenter.cs [hunk 6]
```diff
@@ -5,197 +5,50 @@
...
 
-      public void ValueIsSet(CellValueChangedDTO cellValueChangedDTO)
-      {
-         this.DoWithinLatch(() => AddCommand(_editObservedDataTask.SetValue(_observedData, mapFrom(cellValueChangedDTO))));
-      }
-
-      public void RemoveData(int rowIndex)
-      {
-         if (_observedData.BaseGrid.Count > rowIndex)
-            AddCommand(_editObservedDataTask.RemoveValue(_observedData, rowIndex));
```
Comment (msevestre) [inline src/OSPSuite.Presentation/Presenters/ObservedData/DataRepositoryDataPresenter.cs:66]: Now I think that all the Remove Value etc.. code neecs also to be removed (and the associated command) at the domain level. This is not used anymore. 

  /// <summary>
      /// Command which modifies the units on a specific column of a data repository
      /// </summary>
      /// <param name="dataRepository">The data repository to be modified</param>
      /// <param name="columnId">The column id of the column being modified in the data repository</param>
      /// <param name="newUnit">The new unit</param>
      /// <returns>The command which was used to modify the repositories</returns>
      ICommand SetUnit(DataRepository dataRepository, string columnId, Unit newUnit);

      /// <summary>
      /// Command which removes values from the data repository
      /// </summary>
      /// <param name="observedData">The data repository being modified</param>
      /// <param name="dataRowIndex">The index of the row being removed</param>
      /// <returns>The command which was used to modify the repositories</returns>
      ICommand RemoveValue(DataRepository observedData, int dataRowIndex);

      /// <summary>
      /// Command which adds values to a data repository
      /// </summary>
      /// <param name="observedData">The repository being modified</param>
      /// <param name="dataRowAdded">An entity containing new values for each column of the data repository</param>
      /// <returns>The command which was used to modify the repositories</returns>
      ICommand AddValue(DataRepository observedData, DataRowData dataRowAdded);
/// <summary>
      /// Command which modifies the value of an observed data point in a data repository
      /// </summary>
      /// <param name="observedData">The repository being modified by the command</param>
      /// <param name="cellValueChanged">An entity indicating the new and old cell values as well as the row and column where the data is found</param>
      /// <returns>The command used to modify the repository</returns>
      ICommand SetValue(DataRepository observedData, CellValueChanged cellValueChanged);


And  then the Command, the Parameter CellValueChanged etc..

Comment (msevestre) [reply src/OSPSuite.Presentation/Presenters/ObservedData/DataRepositoryDataPresenter.cs:66]: and I would rename IEditObservedDataTask to IEditObservedDataMetaDataTask or better `IObservedDataMetaDataTask`
as we only have methods related to MetaData

**Review by msevestre — CHANGES_REQUESTED** [review]

File: src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs (hunk 2 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs]: removed

File: src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs (hunk 4 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs]: removed

File: src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs (hunk 5 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.UI/Views/ObservedData/DataRepositoryDataView.cs]: removed

File: src/OSPSuite.Presentation/Presenters/ObservedData/DataRepositoryDataPresenter.cs (hunk 6 above)
Comment (georgeDaskalakis) [reply src/OSPSuite.Presentation/Presenters/ObservedData/DataRepositoryDataPresenter.cs:66]: commands removed and interface renamed

**Comment by georgeDaskalakis** [comment]
@msevestre I think now everything is removed. I kept the context menu text as "Edit...."

**Comment by msevestre** [comment]
FYI @georgeDaskalakis There were still some commands and properties that were not used anymore. Instead of pointing out in the PR, I pushed to your PR so that I can merge.

**Pull request merged**